crates/aptos-logger/src/speculative_log.rs

```diff
@@ -0,0 +1,287 @@
+pub enum LogEntryKind {
+    Execution,
+    Validation,
+    Miscellaneous, // usually testing
+}
+
+#[macro_export]
+macro_rules! speculative_log {
+    ($level:expr, $($args:tt)+) => {
+        if enabled!($level) {
+            speculative_log($level, $($args)+);
+        }
+    };
+}
+
+// Logs directly (no buffering).
+fn log_message(level: Level, context: &AdapterLogSchema, message: &String) {
+    match level {
+        Level::Error => error!(*context, "{}", message),
+        Level::Warn => warn!(*context, "{}", message),
+        Level::Info => info!(*context, "{}", message),
+        Level::Debug => debug!(*context, "{}", message),
+        Level::Trace => trace!(*context, "{}", message),
+    }
+}
+
+/// Adds a message at a specified logging level and given context (that includes txn index)
+/// to the ongoing buffer.
+/// Note: flush_speculativelogs should be called earlier with the appropriate size to
+/// make sure the underlying BUFFERED_LOGS storage is indexed sufficiently.
+pub fn speculative_log(level: Level, context: &AdapterLogSchema, message: String, critical: bool) {
+    let context = context.clone();
+
+    let txn_idx = context.txn_idx;
+    let logs = BUFFERED_LOGS.load();
+    let mut log = logs.0[txn_idx].lock();
+
+    // If speculative_log is called before initialization or somehow without resizing the
+    // buffer properly, log the message directly and log an error. This is just defensive coding
+    // to avoid a panic in such situations (which shouldn't happen).
+    if txn_idx >= logs.0.len() {
+        // Directly log the message.
+        log_message(level, &context, &message);
+        error!(
+            "speculative_log at idx = {}, but Buffer len = {}",
+            txn_idx,
+            logs.0.len()
+        );
+    }
+
+    log.entries.push((level, context, message));
+    if critical {
+        log.critical_err_cnt += 1;
+        logs.1.fetch_add(1, Ordering::SeqCst);
+    }
+}
+
+/// Clears the buffered log for a transaction without logging anything. Useful e.g. when a
+/// speculative transaction execution is aborted by parallel execution (failed validation).
+pub fn clear_speculative_log(txn_idx: usize) {
+    let logs = BUFFERED_LOGS.load();
+    let mut log = logs.0[txn_idx].lock();
+
+    // Logically clear the buffered logs.
+    log.skip_cleared = log.entries.len();
+    logs.1.fetch_sub(log.critical_err_cnt, Ordering::SeqCst);
+    log.critical_err_cnt = 0;
+}
+
+/// Useful for e.g. module r/w fallback to sequential execution, as in this case we would
+/// like to discard all logs coming from the attempted parallel execution.
+pub fn clear_speculative_logs() {
+    // TODO: Could parallelize if needed. Currently does logical clear only.
+    for i in 0..BUFFERED_LOGS.load().0.len() {
+        clear_speculative_log(i);
+    }
+}
+
+/// Resizes the log buffer, as needed (if not large enough) to the new provided size.
+pub fn resize_speculative_logs(num_txns: usize) {
```

Comment:
It's a good question - the simplicity here is that when we have a new block to execute, we can call this, which almost always won't have to resize, and then we can expect to always be within bounds (still defensive coding there to log an error and not panic otherwise, but would be expected behavior to always fit within the size).

However, if we did that on demand, i.e. while logging, it becomes more problematic - threads call VM, VM calls adapter, and we know transaction index and nothing else, we may end up trying to concurrently resize to ever increasing numbers which would be extremely costly.
Even if we knew number of transactions at that point, would have to lock and pretty much do the resize anyway, so thought it's cleaner to require resize upfront, and log an error if you try to log to an index that's not within the size.